### PR TITLE
Richelle (water) - Task List

### PR DESCRIPTION
**Pull request by r-spiel**
# Task List
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe in your own words what the Model is doing in Rails | The model is interacting with the database.  You create the model by running rails generate model (and then adding your database name and you can start with some initial columns).  You then have to use migrations to make any changes to the schema of the database.   
Describe in your own words what the Controller is doing in Rails | The controller is coded in ruby, it gets linked to from the routes which are triggered with the user clicks on things from their browser when they are on your site.  Once they click something (that interacts with your routes), the routes will connect to that method in the Controller which will define what will happen to the database, or what information is being requested etc.  It's then connected to the view (through the routes) to show that info. 
Describe in your own words what the View is doing in Rails | The view controls the HTML that the user will see.  You can insert ruby code to add instances of your active record and get/send info from the database.  
Describe an edge-case controller test you wrote | For the toggle complete tests, I wrote a test to make sure that given an invalid id, it would redirect to the root page.  
What is the purpose of using strong params? (i.e. the params method in the controller) | The strong params are used in the controller to interact with the forms provide a safe way to require and permit data that comes from the forms.  We use strong params within our controllers to provide a safe way to require and permit data that comes from our forms. 
How are Rails migrations related to Rails models? | Migrations are used to add/remove columns and make any other changes (update the name of columns, data types etc.).  
Describe one area of Rails that are still unclear on | I am still confounded by the error messages which I have not had much luck making sense of...  to debug I tend to have to read through each thing that I've changed super detailed for any errors... some error message mage more sense than others, and some error message in the browser are totally unclear to me.

### CODE REVIEW

File: app/views/tasks/index.html.erb [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+            <% if !task.completed_at.nil? %>
+              <br />
+              <%= "Completed: " + task.completed_at %>
+            <% end %>
+          </li>
+        <% end %>
+      </ul>
+    <% else %>
+      <h3>No tasks in the system yet😥 time to add some tasks!</h3>
+    <% end %>
```
Comment (tildeee) [inline app/views/tasks/index.html.erb:48]: This conditional is sooo coooool! I was really happy when I saw there was a message for no tasks. Great touch, I think this is awesome. Keep thinking up good ideas like this!

File: app/controllers/tasks_controller.rb [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+      return
+    end
+
+    @task.destroy
+
+    redirect_to tasks_path
+    return
+  end
+
+  def complete
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:73]: Nitpick: Since this action toggles the `completed_at` property, it might be nice to change the name of this action to something like `toggle`

File: app/controllers/tasks_controller.rb [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    if @task.nil?
+      redirect_to root_path
+      return
+    elsif @task.completed_at.nil?
+      if @task.update(completed_at: Time.now.to_s) # #if true
+        redirect_to tasks_path
+        return
+      else #save failed
+        # #maybe this should be some custom error text to try to complete the task again
+        #render :edit - doing this or something like this does not make sense, there is no form
```
Comment (tildeee) [inline app/controllers/tasks_controller.rb:85]: We'll have better answers for what to do in future projects :)

File: app/views/tasks/index.html.erb [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+          <li>
+            <% if task.completed_at.nil? %>
+              <hr><%= link_to task.name, task_path(task.id) %><%=" - " + task.description %>
+              <br />
+              <div class="task-buttons">
+                <%= button_to "🔲 Mark Complete?", complete_task_path(task.id), method: :patch %>
+              </div>
+            <% else %>
+              <hr><s><%= link_to task.name, task_path(task.id) %><%=" - " + task.description %></s>
+              <br />
```
Comment (tildeee) [inline app/views/tasks/index.html.erb:26]: I'm being _hilariously_ nit-picky here, but `<s>`, and `<br />` both happen to be HTML tags that are _not_ semantic ;) To get a strikethrough, it might be better to use a class with `text-decoration: strikethrough`, and to get a line-break, it might be better to use a combination of margin and display properties.

File: test/controllers/tasks_controller_test.rb [hunk 5]
```diff
@@ -87,38 +83,129 @@
...
   
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+     before do
+        Task.create(name: "Task #1", description: "details for the stuff to do")
+        @task1 = Task.first
+     end
+
+     it "can complete and uncomplete an existing task" do
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:182]: Your indentation is off by one space throughout this whole `describe` block!

File: test/controllers/tasks_controller_test.rb [hunk 6]
```diff
@@ -87,38 +83,129 @@
...
+      must_redirect_to tasks_path
+    end
   end
   
   # Complete for Wave 4
   describe "toggle_complete" do
-    # Your tests go here
+     before do
+        Task.create(name: "Task #1", description: "details for the stuff to do")
+        @task1 = Task.first
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:179]: Nitpick: like how you used variables above, `@task` is a perfectly great variable name; it might be better than `@task1`

File: test/controllers/tasks_controller_test.rb [hunk 7]
```diff
@@ -87,38 +83,129 @@
...
+       }.wont_change "Task.count"
+
+       expect (@task1.reload.completed_at).must_be_kind_of String
+
+       #revert back to nil
+       expect {
+         patch complete_task_path(@task1)
+       }.wont_change "Task.count"
+
+       expect (@task1.reload.completed_at).must_be_nil
```
Comment (tildeee) [inline test/controllers/tasks_controller_test.rb:197]: Nice use of `reload`!

**Review by tildeee — COMMENTED** [review]
# Task List




## Major Learning Goals/Code Review





| Criteria                                           | yes/no, and optionally any details/lines of code to reference |
| -------------------------------------------------- | ------------------------------------------------------------- |
| At least 6 commits with meaningful commit messages | ✔️                                                            |
| Routes follow RESTful conventions                  | ✔️                                                            |
| Uses named routes (like `_path`)                   | ✔️                                                            |
| Creates Models and migrations                      | ✔️                                                            |
| Creates styled views                               | ✔️                                                            |
| Handles errors like nonexistant tasks              | ✔️                                                            |
| Uses `form_with` to render forms in Rails          | ✔️                                                            |

## Functional Requirements/Manual Testing

| Functional Requirement                                                         | yes/no |
| ------------------------------------------------------------------------------ | ------ |
| Successfully handles index & show                                              | ✔️     |
| index & show tests pass                                                        | ✔️     |
| Successfully handles: New, Create                                              | ✔️     |
| New, Create tests pass                                                         | ✔️     |
| Successfully handles: Edit, Update                                             | ✔️     |
| Edit, Update tests pass with valid & invalid task ids                          | ✔️     |
| Successfully handles: Destroy, Task Complete                                   | ✔️     |
| Tests for Destroy & Task Complete include tests for valid and invalid task ids | ✔️     |

## Overall Feedback

| Overall Feedback                | Criteria                                                                                                                                                                                      | yes/no |
| ------------------------------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------ |
| Green (Meets/Exceeds Standards) | 5+ in Code Review && 6+ in Functional Requirements                                                                                                                                            | ✔️
| Yellow (Approaches Standards)   | 3+ in Code Review && 5+ in Functional Requirements, or the instructor judges that this project needs special attention                                                                        |
| Red (Not at Standard)           | 0-2 in Code Review or 0-4 in Functional Reqs, or assignment is breaking/doesn’t run with less than 5 minutes of debugging, or the instructor judges that this project needs special attention |

### Additional Feedback

Richelle!! Great work on this project!! Your controller code looks exactly like how I expect it to look, your tests look great, and your site functions very very very well! I can tell that you put a lot of thought into some decisions, too.

I'm adding a bunch of comments that are a little bit nitpicky on code style-- take these as suggestions for the future!

Overall, your project submission is great; good work, keep it up!

## Code Style Bonus Awards



Was the code particularly impressive in code style for any of these reasons (or more...?)

| Quality              | Yes? |
| -------------------- | ---- |
| Perfect Indentation  |     |
| Elegant/Clever       | ✅    |
| Descriptive/Readable |     |
| Concise              |     |
| Logical/Organized    |     |
